rfcs/0007-api-errors.md

```diff
@@ -0,0 +1,81 @@
+<!-- TOC start (generated with https://github.com/derlin/bitdowntoc) -->
+
+- [SlateDB Error API](#slatedb-error-api)
+   * [Motivation](#motivation)
+   * [Goals](#goals)
+   * [Principles](#principles)
+   * [Public API](#public-api)
+
+<!-- TOC end -->
+
+# SlateDB Error API
+
+Status: Draft
+
+Authors:
+
+* [Jason Gustafson](https://github.com/hachikuji)
+
+## Motivation
+
+Errors exposed through a public API are part of the public API. Users depend on clearly
+documented errors to drive failure handling logic and we cannot change them without considering
+how it affects compatibility. It is important to have common agreement among contributors
+about the guidelines to expose new public errors.
+
+## Goals
+
+The goal of this RFC is to establish a set of principles for exposing public errors
+and a framework for maintaining them.
+
+## Principles
+
+- **Errors should not expose implementation.** The intent is to reserve 
+enough space for the implementation to change without affecting the public API.
+- **Errors should be prescriptive.** Ideally it should be clear to the user what they need to do
+to resolve an issue. For example, we should indicate whether a failure may be transient and
+an operation can be retried. It may not always be possible to have clear guidance for 
+unexpected errors, but ideally we can at least state whether the error is fatal.
+- **Prefer coarse error types.** A useful criteria to decide whether a new error type is needed 
```

Comment:
These all make a ton of sense. I think this is an especially good rule to follow.